onnxruntime/core/graph/graph_utils.h

```diff
@@ -86,20 +84,43 @@ bool GetRepeatedNodeAttributeValues(const Node& node,
   return false;
 }
 
-/** Removes the given Node from the Graph and keeps Graph consistent by rebuilding needed connections.
+/** Check if it will be possible to remove or fuse a node.
     We support the removal of the Node as long as the following conditions hold:
-    - There should be no implicit inputs.
+    - The node should not produce a graph output unless replacement_output_name provides that output.
     - Only one of the outputs is used by downstream operators (but it can have multiple output edges).
-    - If the Node has a single incoming node (and possibly multiple initializers), we can remove the Node and
-      connect its incoming node to its outgoing nodes.
-    - If the Node has a single initializer as input, we remove the Node and feed the initializer as input to its
-      output nodes. */
-bool RemoveNode(Graph& graph, Node& node);
+    - If the Node has a single incoming node, we can remove the Node and connect its incoming node to its 
+      outgoing nodes, if doing so does not clash with any values in any relevant subgraphs. 
+    - If the Node output will be replaced by replacement_output_name, we can remove or fuse the node 
+      if replacement_output_name does not clash with any values in any relevant subgraphs.  
+@param replacement_output_name 
+  If a new NodeArg will be created to replace the node's output (e.g. creating new initializer) 
+  provide the new name that will be used by the NodeArg.
+  If the node is being fused provide the output name from the last node being fused.
+  If nullptr: If the node has one input edge the name from that edge will be used in the checks, 
+              otherwise the node must have one input definition and the name from that will be used in the checks.
+*/
+bool CanRemoveNode(const Graph& graph, const Node& node, const std::string* replacement_output_name = nullptr);
+
+/** Removes the given Node from the Graph and keeps Graph consistent by rebuilding needed connections.
+See CanRemoveNode for details on when removal is allowed.
+@param replacement_output If we are not connecting the single incoming node with the downstream node/s
+                          provide the NodeArg that will replace the output from 'node' and be connected 
+                          with the downstream node/s.
+*/
+bool RemoveNodeAndUpdateEdges(Graph& graph, Node& node, NodeArg* replacement_output = nullptr);
```

Comment:
There is something I dont like about this API.
The replacement_output is confusing, as it is used only for the subcase that the node had only an initializer as input and on top of that, that initializer got replaced.
I would either:
- keep a method with no replacement_output and another that exposes the ReplaceNodeWithNodeArg, or
- add method that replaces an initializer input with another and separately then removes that node -- that would also work right?

Reply:
About the second solution/bullet, maybe the reverse is better: first remove the node and then replace the initializer. This way the remove node and the initializer replacement remain two separate, simpler steps.